pkg/appruntime/chain/mpchain.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2024 KubeAGI.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package chain
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/tmc/langchaingo/chains"
+	"github.com/tmc/langchaingo/llms"
+	"github.com/tmc/langchaingo/prompts"
+	"github.com/tmc/langchaingo/schema"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kubeagi/arcadia/pkg/appruntime/base"
+)
+
+const (
+	// For map-reduce
+	DefaultPromptTemplateForMap = `
+		{{.context}}
+
+		With above content, please summarize it with only half content size of it.
+		`
+	DefaultPromptTemplatForReduce = `"{{.context}}"`
+
+	// For post process the map-reduced summary
+	DefaultPromptTemplateForPostMapReduce = `
+		Here is the map-reduced summary of a document:
+
+		Summary: {{.summary}}
+
+		Now please answer the following question based on the above document summary. Make sure the answer is using same language with the question:
+
+		Question: {{.question}}
+
+		Answer:
+	`
+
+	DefaultSummaryMaxNumberOfConcurrent = 2
+	DefaultDocumentChunkSize            = 1024
+	DefaultDocumentChunkOverlap         = 100
+)
+
+type MapReduceChain struct {
```

Comment:
This `MapReduceChain `  can be used as a extension of `LLMChain` or `RetrivalQAChain` to handle documents.